### DISCUSSION TITLE
Unclear on intended behavior for lazy compilation of entrypoints

### DISCUSSION THREAD

**Discussion by karmeleon** [webpack/webpack#14619] [Q&A]
My team has a project using webpack-dev-server that hosts dozens of entrypoints. Compiling them all at once takes several minutes, so we set up a system that reads a list of entrypoint names from a config file and only compiles the entrypoints that a dev is actually interested in. This cuts down compile time significantly, but this behavior can trip up devs who aren't aware of exactly how it works.

Webpack's new lazy entrypoint experiment seemed like a perfect solution. My understanding was that entrypoints would be compiled on-demand, and then loaded onto the page to work as if the entrypoints had been pre-compiled. However, once I got it set up, I found that Webpack would correctly lazily compile entrypoints as they were requested, but would not load them onto the page until the page was refreshed.

Is this behavior intentional? If so, what is the recommended way to serve apps with lazily-compiled entrypoints?

**Comment by alexander-akait** [comment]
Please provide example of the problem

**Reply by karmeleon** [reply]
I made a minimal repo that uses this feature and it worked perfectly -- the entrypoint was built lazily, then loaded onto the page and executed when it was ready. The feature clearly does work as I expected, but there's something wrong with my usage of it causing it to malfunction.

Digging a bit more into my repo, I found that the issue was that the main app entrypoint was being compiled lazily before the HMR bundle was on the page. This means the HMR client missed the "bundle finished compiling" signal, so never queried the dev server for the new bundles. I used the `test` option to exclude all HMR-related bundles from lazy compilation and in my testing this seems to have fixed the issue, though I have a hunch I just worked around some deeper root cause.

Unfortunately since the system I'm working on is so complicated I haven't been able to produce a minimum reproduction case yet, but I'll update if I can find anything else that points towards an underlying cause.

**Reply by tmeasday** [reply]
I'm not sure if it is the same issue or not, but if you are using [`webpack-hot-middleware`](https://github.com/webpack-contrib/webpack-hot-middleware) (which gets you to add an entry point `webpack-hot-middleware/client`) you need to either disable entry point lazy compilation (`lazyCompilation: { entries: false }`) or filter it out as you have, @karmeleon which may explain why you had trouble reproducing.

This is easy to reproduce (I can submit a repro if it helps), although I am not sure if technically it should be considered a bug. I suspect that should be documented alongside the lazy compilation docs @alexander-akait as I suppose it is a pretty common setup (using `webpack-hot-middleware`).

**Reply by sokra** [reply]
Lazy compilation uses HMR to update the page after a lazy compilation has happened (technically you could also reload manually but that's not that smooth).
So you need to exclude all HMR related entrypoints from being lazy compiled. Some well known are already excluded. I'm happy to accept a PR with the webpack-hot-middleware entry too, if this is missing.

You will know if that is the problem when it fixes itself by manually reloading.

**Reply by tmeasday** [reply]
I'm planning on submitting a PR for this tomorrow (my time). Thanks @sokra
